### PR TITLE
Usability Updates for StdChains

### PR DESCRIPTION
**Pull request by hexonaut**
Ping @mds1

Two useful updates to the `StdChains`.

1. Adds the `chainAlias` to the`Chain` struct. We need access to this as the chain alias is useful for defining chain-specific stuff inside configuration json files such as:

```
{
    "domains": {
        "mainnet": {
            "chainlog": "0xdA0Ab1e0017DEbCd72Be8599041a2aa3bA7e740F"
        },
        "optimism": {
            "l1Messenger": "0x25ace71c97B33Cc4729CF772ae268934F7ab5fA1",
            "l2Messenger": "0x4200000000000000000000000000000000000007"
        },
        "arbitrum_one": {
            "inbox": "0x4Dbd4fc535Ac27206064B68FfCf827b0A60BAB3f",
            "arbSys": "0x0000000000000000000000000000000000000064"
        }
    }
}
```

This is nicer to define than using chain ids.

2. Allow automatic environment variable overrides via `ALIAS_RPC_URL`. For example a user can define `MAINNET_RPC_URL` to set the mainnet rpc url.

This saves having to repeat this pattern in all `foundry.toml` which will be common for us at least:

```
[rpc_endpoints]
mainnet = "${MAINNET_RPC_URL}"
goerli = "${GOERLI_RPC_URL}"
optimism = "${OPTIMISM_RPC_URL}"
... etc
```

### CODE REVIEW

File: src/StdChains.sol
Comment (iamchrissmith) [inline src/StdChains.sol:126]:
```suggestion
        bytes memory strb = bytes(str);
        bytes memory copy = new bytes(strb.length);
```

File: src/StdChains.sol
Comment (iamchrissmith) [inline src/StdChains.sol:148]:
```suggestion
                chain.rpcUrl =
                    vm.envOr(string(abi.encodePacked("RPC_URL_", _toUpper(chainAlias))), defaultRpcUrls[chainAlias]);
```

File: src/StdChains.sol
Comment (iamchrissmith) [inline src/StdChains.sol:192]:
```suggestion
        setChainWithDefaultRpcUrl(
            "bnb_smart_chain", Chain("BNB Smart Chain", 56, "", "https://bsc-dataseed1.binance.org")
        );
```

File: src/StdChains.sol
Comment (iamchrissmith) [inline src/StdChains.sol:187]:
```suggestion
        setChainWithDefaultRpcUrl(
            "polygon_mumbai", Chain("Polygon Mumbai", 80001, "", "https://rpc-mumbai.maticvigil.com")
        );
```

**Comment by iamchrissmith** [comment]
one last format error that is showing up locally, but GH won't let me suggest a change:

the formatter wants this function declaration https://github.com/foundry-rs/forge-std/pull/268/files#diff-e21e7255e73417cde7931c5f5ffbe44f1009bbe12ccdf684942299c83448f7ddR140-R143

to be one line:

```solidity
function getChainWithUpdatedRpcUrl(string memory chainAlias, Chain memory chain) private returns (Chain memory) {
```

**Comment by hexonaut** [comment]
Fixed all formatting.

File: src/StdChains.sol
Comment (mds1) [inline src/StdChains.sol:124]: Do we have a source we can link to for where this came from originally? (assuming you didn't write it just for this PR, but maybe you did)

File: src/StdChains.sol
Comment (mds1) [inline src/StdChains.sol:146]: Thoughts on `ALIAS_RPC_URL` instead of `RPC_URL_ALIAS`? I'm personally ok with either, but from a quick search it seems the former is a much more common naming scheme—1.1k results for the [former](https://sourcegraph.com/search?q=context:global+MAINNET_RPC_URL+count:5000&patternType=standard&sm=1), 101 for the [latter](https://sourcegraph.com/search?q=context:global+RPC_URL_MAINNET+count:5000&patternType=standard&sm=1)

image
image

File: src/StdChains.sol
Comment (mds1) [inline src/StdChains.sol:167]: Did anything change in this section aside from adding the `""` throughout for chain alias? I don't think so but the diff came out annoying so just checking 😅

File: src/StdChains.sol
Comment (mds1) [inline src/StdChains.sol:47]: Nit, may as well as add a bit more detail
```suggestion
        // The chain's alias (i.e. what gets specified in `foundry.toml`).
```

**Review by mds1 — COMMENTED** [review]
I like this idea, thanks a lot! @ZeroEkkusu you led a lot of the stdChains work so would love for you to take a look if you have time

File: src/StdChains.sol
Comment (mds1) [reply src/StdChains.sol:167]: Also this is a breaking change to the interface which I'd like to avoid if we can. What do you think of creating a new struct called `ChainData` and using that as the chain return data and keeping `Chain` as the input data?

File: src/StdChains.sol
Comment (hexonaut) [reply src/StdChains.sol:124]: I wrote it. :)

File: src/StdChains.sol
Comment (hexonaut) [reply src/StdChains.sol:146]: I'm okay with switching the order. I suppose putting the alias in front matches the style of `ETH_RPC_URL` that dapp tools defined.

Comment (hexonaut) [reply src/StdChains.sol]: done https://github.com/foundry-rs/forge-std/pull/268/commits/758fd9066dca595a71440c7942e8164d4b4c473c

File: src/StdChains.sol
Comment (hexonaut) [reply src/StdChains.sol]: done https://github.com/foundry-rs/forge-std/pull/268/commits/6ac3ebb618a2ea239860aeb45bddf506e6ec8ff7

File: src/StdChains.sol
Comment (hexonaut) [reply src/StdChains.sol:167]: > Did anything change in this section aside from adding the `""` throughout for chain alias? I don't think so but the diff came out annoying so just checking 😅

Yes this is just the `""` addition.

> Also this is a breaking change to the interface which I'd like to avoid if we can. What do you think of creating a new struct called `ChainData` and using that as the chain return data and keeping `Chain` as the input data?

Changing the output is also a breaking change. Does it make more sense to break the input or the output? If I had to guess breaking the output would actually be a larger breaking change as this is people defining custom chains vs using the common ones defined.

I can proceed if you think this is a better way to go, but wanted to double check.

**Review by ZeroEkkusu — COMMENTED** [review]
Is the point of change 1 to be able to get chainAlias from a Chain? If so, what if instead of a breaking change, you add getAlias function that takes a Chain as a parameter, uses its id to get a Chain from getChain and then compares their hashes to see if they’re equal. Is so, it returns the alias from idToAlias. Otherwise, reverts.

**Comment by ZeroEkkusu** [comment]
You have more insight, since you’re using it.

One thing I thought about before is whether anyone uses `Chain.name`. It seemed more useful to be `Chain(“chain_alias”, 123, “https://…”)`.

You should go with what’s more practical in general.

**Comment by hexonaut** [comment]
That's a good point. Other than maybe log output I don't think the human-friendly name is all that useful, and even if it is it's not hard for users to define this outside of `StdChains`. Perhaps we can just use the chain alias in the name so things don't break?

**Comment by ZeroEkkusu** [comment]
I think it’s the best solution DX wise, for example, you won’t need to pass a key and Chain in setChain anymore, only a Chain.

This solution is also a breaking change.

**Comment by hexonaut** [comment]
We could also just leave the interface and by convention use the chain alias for the name for the defaults.

**Comment by ZeroEkkusu** [comment]
There are quite a few options - I’ll let @mds1 and you decide, and pitch in if necessary.

**Comment by mds1** [comment]
Thanks for the discussion here @hexonaut and @ZeroEkkusu. One other thought is that if we make this a breaking change to decouple the `Chain` input struct and the `ChainData` return struct (or whatever names we go with), that does set us up to expand the returned `ChainData` in the future, which could be useful. For example, you could imagine adding a `bool supportsEip1559` flag or something.

So perhaps we just go that route with the breaking change? Was hoping to avoid breaking changes but since we can't (and I agree having alias as a field makes sense), we may as well go with the one that sets us up best going forward. Which I think is adding a new struct so `setChain` can stay the same for now, and we have a breaking change on the return data struct to add the `alias`.

**Comment by hexonaut** [comment]
Ok sounds good. Will make this change in a moment.

File: src/StdChains.sol
Comment (mds1) [inline src/StdChains.sol:42]: Nit, wdyt of switching the names of the `Chain` and `ChainData` structs? `ChainData` sounds like broader scope and the return data has more info. Def open to keeping this as-is though too. Otherwise this PR LGTM

Comment (mds1) [reply src/StdChains.sol:42]: Ah just saw [your comment](https://github.com/foundry-rs/forge-std/pull/268#issuecomment-1375961973), ok that works for me

**Comment by mds1** [comment]
Just needs `forge fmt`, @ZeroEkkusu do you want to take a quick look also before merging?

**Review by mds1 — APPROVED** [review]
Thank you!

File: src/StdChains.sol
Comment (ZeroEkkusu) [inline src/StdChains.sol:101]: Can you update these error messages to say `ChainData`, and the description on top of the file?

Also, do you think it's useful, to be able to "convert" `Chain` into `ChainData` so you can pass it to `setChain`, by adding an alternative signature:

```
chain = getChain(...)
chain.name = ...
setChain(..., chain)
```

Otherwise, LGTM, @mds1 can merge when ready.

Comment (hexonaut) [reply src/StdChains.sol]: Done. https://github.com/foundry-rs/forge-std/pull/268/commits/43ff37017acaa235d5c99d42b73b7fa665a530df

**Review by ZeroEkkusu — APPROVED** [review]

File: src/StdChains.sol
Comment (ZeroEkkusu) [inline src/StdChains.sol:134]: We can require that `chain.chainAlias` is empty or equal to `chainAlias`. Otherwise, revert with `StdChains setChain(string,Chain): Will not convert Chain to ChainData. chainAlias is "abc", but Chain.chainAlias is "xyz". Make sure you are passing the right Chain.`.

Comment (hexonaut) [reply src/StdChains.sol:134]: I'm okay with either. @mds1 lmk your thoughts.

Comment (mds1) [reply src/StdChains.sol:134]: Right now `setChain` takes `ChainData` as input which doesn't have the `chainAlias` field. Then we set the chain alias in storage for the full `Chain` struct. So the current approach seems good to me unless I'm misunderstanding something?

Comment (ZeroEkkusu) [reply src/StdChains.sol:134]: It’s to protect users from accidentally pressing a wrong chain. Because this function reduces Chain to ChainData, it is possible to do setChain(“chain_1”, chain2). If the intention is not to assign “chain_1” to chain2, the function should let the users now of a possible error. In practice, it should happen rarely.

If you don’t think it should be there, you can merge the PR.

Comment (ZeroEkkusu) [reply src/StdChains.sol:134]: @mds1  Note that this is a new function to let you pass a Chain instead of ChainData for the ease of use. This is since Chain has enough data to satisfy the ChainData requirement (but not the other way around).

Comment (mds1) [reply src/StdChains.sol:134]: Another idea is just `setChain(Chain memory chain)` and removing the separate `chainAlias` parameter. Wdyt of merging this to avoid churn, and we can follow up in a separate PR to consider implementing that overload? Since that shouldn't be a breaking change

Comment (ZeroEkkusu) [reply src/StdChains.sol:134]: If you think so, yes - I don’t want to block the PR. The overload is already implemented, btw.

Re setChain(Chain): It’s possible that chain.chainAlias is not assigned; that would need to be checked. The original idea was just to not need to convert Chain to ChainData manually, if you have a Chain. The same signature, so things are simple. But, yes, there could be an alternative way of setting a Chain (this one).

Comment (mds1) [reply src/StdChains.sol:134]: Ah whoops yes it's already implemented, sorry 😅 

I'm going to merge this as is since CI passed, and @hexonaut or @ZeroEkkusu if either of you wants to open a follow up PR with the above changes to make setting safer, that should be a quick one to merge. Thanks all!

**Pull request merged**
